File: algorithms/java/99_recover_binary_search_tree/Recover.java
class Recover {
    class State {
        TreeNode prev = null;
        TreeNode node1 = null;
        TreeNode node2 = null;
    }

    public void recoverTree(TreeNode root) {
        if (root == null) return;
        State state = new State();
        inorder(root, state);
        swap(state.node1, state.node2);
    }
    private void inorder(TreeNode root, State state) {
        if (root == null) return;
        inorder(root.left, state);
        if (state.prev != null && state.prev.val > root.val) {
            if (state.node1 != null) {
                state.node1 = state.prev;
            }
            state.node2 = root;
        }
        state.prev = root;
        inorder(root.right, state);
    }
    private void swap(TreeNode n1, TreeNode n2) {
        int tmp = n1.val;
        n1.val = n2.val;
        n2.val = tmp;
    }
}

File: algorithms/cpp/99_recover_bst/Recover.cpp
class Recover {
private:
	struct State {
        TreeNode* prev;
        TreeNode* node1;
        TreeNode* node2;
        State() : prev(nullptr), node1(nullptr), node2(nullptr) {}
    };

    void inorder(TreeNode* root, State* state) {
        if (root == NULL) {
            return;
        }
        inorder(root->left, state);
        if (state->prev != NULL && state->prev->val > root->val) {
            if (state->node1 == NULL) {
                state->node1 = state->prev;
            }
            state->node2 = root;
        }
        state->prev = root;
        inorder(root->right, state);
    }
public:
	void recoverTree(TreeNode* root) {
        State state;
        inorder(root, &state);
        swap(state.node1->val, state.node2->val);
    }
};
